### PR TITLE
Dhanisha

### PR DESCRIPTION
**Pull request by DhanishaHarshad**
Summary
Implements a FastAPI-based insurance discount calculator following Test-Driven Development (TDD) methodology. The API calculates discount rates based on driver age and experience with comprehensive validation and error handling.
Business Logic

Base discount: 0%
Age 25+: +5% discount
Experience 5+ years: +5% discount
Age 40+: Additional +5% discount
Experience 10+ years: Additional +5% discount
Maximum cap: 20% total discount

Key Features

Input validation: Prevents negative values and impossible experience scenarios (experience > age)
HTTP status codes: Returns 400 for business logic errors, 422 for format/type validation errors
Comprehensive error handling: Detailed error messages for various failure scenarios
API documentation: OpenAPI/Swagger documentation with examples

Technical Implementation

Framework: FastAPI with Pydantic models
Validation: Type checking and business rule enforcement
Testing: 57 comprehensive test cases covering all scenarios
Architecture: Separation of business logic from API layer

Test Coverage
Implemented 30 distinct test scenarios including:

Boundary conditions (exact age/experience thresholds)
Edge cases (zero values, maximum values)
Error scenarios (negative inputs, invalid data types)
Realistic driver profiles
API integration testing

Files Added/Modified

discount_logic.py - Core business logic implementation
main.py - FastAPI application with endpoint definition
test_all_thirty_scenarios.py - Comprehensive TDD test suite
test_discount.py - Additional unit and integration tests

Testing Results

Total tests: 57
Passing: 57 (100%)
Coverage: All business rules and edge cases

API Endpoints

POST /discount - Calculate discount rate
GET / - API information
GET /health - Health check

### CODE REVIEW

File: test_all_thirty_scenarios.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        response = client.post("/discount", json={"age": True, "experience": False})
+        # FastAPI converts True to 1, False to 0, so this might be valid
+        if response.status_code == 200:
+            # If accepted, should return 0% discount for age=1, experience=0
+            assert response.json()["discount_rate"] == 0
+        else:
+            # If rejected, should be 422 for type validation
+            assert response.status_code == 422
+    
+    def test_api_case_25_array_values(self):
```
Comment (AndyGuffey) [inline test_all_thirty_scenarios.py:210]: Looks good, But could possibly have some inproved comments for clarity!

File: test_all_thirty_scenarios.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        """API Test 25: Array values → 422 error"""
+        response = client.post("/discount", json={"age": [30], "experience": [5]})
+        assert response.status_code == 422
+
+    # ADDITIONAL BOUNDARY TESTS
+    def test_api_multiple_boundary_scenarios(self):
```
Comment (AndyGuffey) [inline test_all_thirty_scenarios.py:216]: Looks very good !

**Review by AndyGuffey — COMMENTED** [review]
Looks very good, possibly could have improved comments for the 30 test scenarios

**Review by AndyGuffey — APPROVED** [review]

**Review by AndyGuffey — APPROVED** [review]

**Review by AndyGuffey — APPROVED** [review]

**Review by AndyGuffey — APPROVED** [review]

**Review by AndyGuffey — APPROVED** [review]
Didn't See any merge conflicts, looks really good!
